File: docs/get-started/python-quickstart/code/part-2/src/summarize/gateway.py
# @@@SNIPSTART python-quisckstart-part2-gateway-py
from resonate.scheduler import Scheduler
from resonate.storage import RemotePromiseStore
from fastapi import FastAPI
from pydantic import BaseModel
from summarize.app import downloadAndSummarize
import uvicorn

app = FastAPI()

class SummarizeRequest(BaseModel):
    url: str

class SummarizeResponse(BaseModel):
    summary: str
    error: str = None

# Define a route handler for the /summarize endpoint
@app.post('/summarize')
def summarize_route_handler(request: SummarizeRequest) -> SummarizeResponse:
    try:
        # Extract the URL from the request
        url = request.url
        # Summarize the text
        promise =  resonate.run(f"summarize-{url}", downloadAndSummarize, url=url)
        return {'summary': promise.result()}
    except Exception as e:
        return {'error': str(e)}

# Create a Resonate Scheduler   
resonate = Scheduler(durable_promise_storage=RemotePromiseStore(url="http://localhost:8001"))
# Register the downloadAndSummarize function with the Resonate scheduler
resonate.register(downloadAndSummarize)

# Define a main function to start the FastAPI app
def main():
    uvicorn.run(app, host="127.0.0.1", port=5000)
    print("Serving HTTP on port 5000...")

# Run the main function when the script is executed
if __name__ == '__main__':
    main()
# @@@SNIPEND

File: docs/get-started/typescript-quickstart/code/part-2/src/gateway.ts
import { Resonate } from "@resonatehq/sdk";
import express, { Request, Response } from "express";
import { downloadAndSummarize } from "./app";

// @@@SNIPSTART quickstart-ts-part-2-server-url
// Initialize a Resonate application.
const resonate = new Resonate({ url: "http://localhost:8001" });
// @@@SNIPEND

// @@@SNIPSTART quickstart-ts-part-2-register-timeout
// Register a function as a Resonate function
resonate.register(
  "downloadAndSummarize", // function name
  downloadAndSummarize, // function pointer
  resonate.options({ timeout: 60000 }) // set a total execution timeout of 1 minute
);
// @@@SNIPEND

// Start the Resonate application
resonate.start();

// Initialize an Express application.
const app = express().use(express.json());

// Register a function as an Express endpoint
app.post("/summarize", async (req: Request, res: Response) => {
  const url = req.body?.url;
  try {
    // Call the resonate function
    let summary = await resonate.run(
      "downloadAndSummarize", // function name
      `summarize-${url}`, // unique identifier of the promise
      url // function argument
    );
    res.send(summary);
  } catch (e) {
    res.status(500).send("An error occurred.");
  }
});

// Start the Express application
app.listen(3000, () => {
  console.log("Listening on port 3000");
});
